vllm/model_executor/layers/fused_moe/layer.py

```diff
@@ -645,10 +684,32 @@ def select_experts(hidden_states: torch.Tensor,
 
         return topk_weights, topk_ids
 
+    def naive_multicast(self, x: torch.Tensor, max_num_tokens: int):
+        assert (len(x.shape) == 2)
+        num_tokens = x.size(0)
+        buffer = torch.zeros((self.dp_size, max_num_tokens, x.size(1)),
+                             device=x.device,
+                             dtype=x.dtype)
+
+        buffer[self.dp_rank, :num_tokens, :].copy_(x)
+
+        x = get_dp_group().all_reduce(buffer)
+        x = x.view(-1, x.size(-1))
+        return x
+
     def forward(self, hidden_states: torch.Tensor,
                 router_logits: torch.Tensor):
         assert self.quant_method is not None
 
+        if self.dp_size > 1:
+            num_tokens_across_dp = get_forward_context().num_tokens_across_dp
+            max_num_tokens = max(num_tokens_across_dp)
+            num_tokens = hidden_states.size(0)
+
+            assert num_tokens_across_dp is not None
+            hidden_states = self.naive_multicast(hidden_states, max_num_tokens)
```

Comment:
one suggestion (not sure if it sounds better): we can align with the terminology from deepep, use `dispatch` and `combine` here. and right now `dispatch` actually gathers all tokens, while `combine` is just a slicing op.

Reply:
I'm a little opposed to this because the operations here work sufficiently differently compared to `dispatch` and `combine` that I'm not sure it makes sense to share names. I might change my mind on it though.

Reply:
sounds good.